docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+
+
+Corner cases / Open questions
+-----------------------------
+
+There are some corner cases that may have interesting implications here, in particular when objects are passed between modules.
+For example, consider a package `A` that exposes the following class (straw-man syntax for declaring versions, probably not what we want to end up with):
+
+```verona
+class A
+{
+	// This is always public
+	isSomething() : Bool;
+	// This is public in any API version >= 2.0
+	[[version=2.0+]]
+	isSomethingElse() : Bool;
+}
+```
+
+If a file in another package imports this then it may do something like:
+
+```verona
+import "A@1.0" as A;
+
+interface IsSomethingElse
+{
+  isSomethingElse() : Bool;
+};
+
+var a = A.A();
+```
+
+The variable `a` here does not implement the interface `IsSomethingElse` because the `isSomethingElse` method is not visible here.
+If `a` is passed to another package containing a similar interface, should it match?
+There are two obvious answers:
+
+ - If the other package imports `A` with a version <2.0, then it should not see the method.
+ - If the other package imports `A` with a version >=2.0, then it should see the method.
+
+The third case is the most interesting.
+What is the correct behaviour if the other package does not import `A` at all?
+Note that, without full dataflow analysis, it is not possible to tell in the general case which package the object was allocated in (for example, it might be stored in an array of `Any` allocated by a fourth module and populated by many other things in the system) and so the question reduces to selector identity.
+Specifically: Is the selector `isSomethingElse/1` in a module that does not import `A` the same as the selector `isSomethingElse@2.0/1` in `A`?
+There are at least two possible answers here:
+
+ - If a module does not directly import another, it sees all versioned symbols as if they were public.
+ - If a module does not directly import another, it does not see any versioned symbols.
+
+The second option is almost certainly too limiting because it effectively prevents any reliance on structural typing across packages if versioning is used.
+The first option potentially introduces breaks but is likely to avoid most of the common causes of breakage because the code in the module that is receiving objects via interface types and which cannot explicitly name the concrete types that they represent must be robust in the presence of new concrete types.
+This is more obvious when you consider the transitive case:
+
+```verona
+// In Module A:
+class A
+{
+  ...
+  [[version=1.1+]]
+  foo(Self & mut)
+};
+
+// In module B:
+interface Fooable
+{
+  foo(Self & mut);
+}
+
+interface Any {}
+
+getFooable(a : Any & mut) : Fooable | None
+{
+  match (a) {
+    Fooable => a;
+    _ => None
+  }
+}
```

Comment:
I recognize that this is probably just the Haskell programmer in me speaking, but this really feels like uncomfortably excessive duck typing.  I think I would rather have a first-class notion of selectors, so that `B`'s `Fooable::foo` really was imported from `A`, at a particular version, at that.

The downside of of this kind of approach, which is more or less what Haskell does, is that it becomes impossible to name a type without picking up a dependency on the defining package.  The ecosystem has responded by either using exclusively compositions of `base` types (eschewing `newtype` or `data` definitions entirely, e.g., in the case of `lens`), or by splitting packages into what amount to interface, usually dependency-less packages (`foo-types`) and implementation, dependency-carrying packages (`foo`).

Reply:
David suggested in our 1:1 today that, rather than making selectors generative (and so, nominal) rather than applicative (to borrow terminology from ML, I think?), we could have a generalized "selector attribute" mechanism, with generative (nominal) attributes.  Specifically,
* Concrete types continue to define (at most) one implementation of a given method name and arity (or name and argument type vector?) and may stick a set of named attributes on this definition.
* Interface definitions continue to enumerate entries as now, but they may additionally specify constraints on any would-be matching selector.

This generalizes the version scheme, in that versions are now just generic attributes on concrete types' methods (that is, each method is annotated with all the versions it was present in) and interfaces require the appropriate version label.  It also serves as a kind of nominal proxy for semantics: if David and I have different ideas about what kind of thing goes inside the result of `std::string serialize(Self)` (say), then I can annotate my methods with an annotation from a package I control and David his with one he controls and there won't be risk of confusion.

Reply:
To add a bit more detail:

We discussed Wes' solution a bit.  The core problem is that it doesn't allow defining open-world interface types (you must specify the concrete types that expose selectors, which means that you're defining a closed-world set and at that point you may as well just use union types).  

We discussed a bit the way that nominal typing is used in Smalltalk-family languages (Java, Objective-C) as a proxy for semantics.  I recalled a random ArXiv paper I read about 15 years ago from someone reimagining Objective-C by providing a central table of selectors that came with contracts for the implementation and using that, rather than simple name-based lookup, for dispatch (I don't think it was ever published anywhere and I never heard anything about the project afterwards).  This felt a bit closer to what I wanted than nominal typing.  

The suggestion that we came up with was to introduce explicit annotations on selectors (which would be first-class concrete types, so are namespaced and exported from individual packages not just ad-hoc strings), with the ability to express a set of them on concrete implementations and refer to them in interface definitions with some simple set-theory operations.  I think this gives me the things I want from nominal typing (where I don't actually want nominal typing, it's just the thing I often have), without introducing a parallel nominal type system (everything is still structural, it's just that now you have a way of unambiguously expressing intentionality in the structural type system). 

I *think* the versioning information could *almost* be treated as a special case if we had this general mechanism, but we should probably discuss them separately and if we want both then consider how they can be merged later.